Source/WebKit/UIProcess/WebFrameProxy.cpp

```diff
@@ -428,14 +428,20 @@ void WebFrameProxy::prepareForProvisionalLoadInProcess(WebProcessProxy& process,
     if (isMainFrame())
         return completionHandler();
 
-    Site navigationSite(navigation.currentRequest().url());
     RefPtr page = m_page.get();
+    Site navigationSite(navigation.currentRequest().url());
     // FIXME: Main resource (of main or subframe) request redirects should go straight from the network to UI process so we don't need to make the processes for each domain in a redirect chain. <rdar://116202119>
     RegistrableDomain mainFrameDomain(page->mainFrame()->url());
 
     m_provisionalFrame = nullptr;
     m_provisionalFrame = makeUnique<ProvisionalFrameProxy>(*this, group.ensureProcessForSite(navigationSite, process, page->preferences()));
-    page->websiteDataStore().protectedNetworkProcess()->addAllowedFirstPartyForCookies(process, mainFrameDomain, LoadedWebArchive::No, WTFMove(completionHandler));
+    protectedProcess()->sendWithAsyncReply(Messages::WebPage::GetFrameOwnerPermissionsPolicy(m_frameID), [weakThis = WeakPtr { *this }, page = m_page, process = Ref { process }, frameID = m_frameID, mainFrameDomain = WTFMove(mainFrameDomain), completionHandler = WTFMove(completionHandler)](auto policy) mutable {
+        if (!page)
+            return completionHandler();
+
+        process->send(Messages::WebPage::SetFrameOwnerPermissionsPolicy(frameID, policy), page->webPageIDInProcess(process));
+        page->websiteDataStore().protectedNetworkProcess()->addAllowedFirstPartyForCookies(process, mainFrameDomain, LoadedWebArchive::No, WTFMove(completionHandler));
+    }, page->webPageIDInMainFrameProcess());
```

Comment:
I am not sure we need to send permission policy at the time we prepare the provisional load.
We only need to send it when starting to process the response data.
Maybe we should send it at the time the UIProcess receives the response policy check?

If the policies are not too big, maybe we could send them as part of the decide for policy action check, like we do for other iframe states?

Reply:
The latest patch includes the owner policy in navigation request instead; by doing that we don't need independent messages to set and update owner's permissions policy in subframe processes.